### ISSUE TITLE
relaxing a non-`Sized` bounds is a hard-warning

### ISSUE THREAD

**Issue by lcnr** [rust-lang/rust#135809]
https://github.com/rust-lang/rust/blob/b605c65b6eb5fa71783f8e26df69975f9f1680ee/compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs#L97-L111

```rust
fn foo() -> impl ?Send {}
```

```
warning: relaxing a default bound only does something for `?Sized`; all other traits are not bound by default
 --> src/lib.rs:1:18
  |
1 | fn foo() -> impl ?Send {}
  |                  ^^^^^
```

This should either be a (future-compat) lint or a hard-error.

**Comment by lcnr** [comment]
this warning exists since pre 1.0 🤔

**Comment by theemathas** [comment]
This code compiles. Is it supposed to?

```rust
trait Generic<T> {}
trait Huh {
    type Assoc;
}
pub fn f<T: ?Generic<<T as Huh>::Assoc>>() {}
```

**Comment by theemathas** [comment]
Clearly this isn't supposed to compile, right?

```rust
trait Generic<const N: usize> {}

pub fn f<T: ?Generic<{ panic!() }>>() {}
```

**Comment by lcnr** [comment]
> This code compiles. Is it supposed to?

> Clearly this isn't supposed to compile, right?

I don't think they should, which is why I opened this issue. We should move all of these to a hard error
